### DISCUSSION TITLE
Query that finds matches, but excludes descendants that also match?

### DISCUSSION THREAD

**Discussion by dhorkin** [tree-sitter/tree-sitter#2369] [Q&A]
I'm currently using `tree-sitter-heex` as my language. This is my query:
```
[
    (
      (attribute
          (attribute_name) @attribute-name
          (quoted_attribute_value
              (attribute_value) @attribute-value
          )
      )
      (#eq? @attribute-name "data-test")
    )
    (
      directive
        (expression_value) @expression
        (#match? @expression "data-test")
    )
]
```

When I run the above query over these contents:
```html
<div data-test='a'>
    <div data-test='b'></div>
    <div data-test='c'>
        <div data-test='d'></div>
    </div>
</div>
```
it correctly pulls out `['a', 'b', 'c', 'd']`

However, my ultimate goal is to maintain their relative positions, i.e.:
```
{
    a: {
        b: null,
        c: {
             d: null
        }
    }
}
```

Since I don't think there's a way in tree-sitter to generate that directly, I'm first getting the flat list, then re-running the query on that node:
```js
// [a, b, c, d]
const captures = query.captures(tree.rootNode)
    .filter((match) => match.name === 'attribute-value' || match.name === 'expression' && match.node.text.includes('data-test'))l

const a = captures[0].node; // simplification (needed to call .parent a few times)
// [b, c, d]
const aDescendantMatches = query.captures(a)
  .filter((match) => match.name === 'attribute-value' || match.name === 'expression' && match.node.text.includes('data-test'));
```

The problem with this approach is `aDescendantMatches` should include `b` and `c`, but not `d`, and I'm not sure how to modify the query appropriately.

If there's a better way to build my map than something like this, feel free to show how to do that instead.

Thanks!

**Comment by dhorkin** [comment]
This is very similar, but not quite the same: https://github.com/tree-sitter/tree-sitter/discussions/2070

I'm trying to use the advice in there (using . and _) to modify my query, but I'm clearly not doing something right:

```
      (_ . (_ . [
        (
          (attribute
              (attribute_name) @attribute-name
              (quoted_attribute_value
                  (attribute_value) @attribute-value
              )
          )
          (#eq? @attribute-name "data-test")
        )
        (
          directive
            (expression_value) @expression
            (#match? @expression "data-test")
        )
      ].)  )
```

still produces [b, c, d]

Various permutations either result in nothing being found or the same results. Not sure what I'm doing wrong here.

**Comment by dhorkin** [comment]
I've partially figured out my issue - it appears the trick in #2070 only works if your inner query has enough layers. So the reason it's not working for me is I'm directly trying to match attribute, and not the thing that might contain the attribute. Since an attribute could be the child of any number of layers, i'm trying to avoid explicitly putting every permutation that could contain an attribute. 

Still trying to work on that, but I'm getting there and will post a solution when i fully get it.

**Comment by dhorkin** [comment]
It looks like the basic issue I'm having is that the attribute tag is nested with other tags, and as far as tree-sitter is concerned, the node in question is a sibling, rather than a descendant.

This:
```html
<div data-test='a'>
    <div data-test='b'></div>
    <div data-test='c'>
        <div data-test='d'></div>
    </div>
</div>
```

Produces:
>(fragment 
    (tag 
        (start_tag (tag_name) (attribute (attribute_name) (quoted_attribute_value (**_attribute_value_**))))
        (tag (start_tag (tag_name) (attribute (attribute_name) (quoted_attribute_value (**_attribute_value_**)))) (end_tag (tag_name)))
        (tag (start_tag (tag_name) (attribute (attribute_name) (quoted_attribute_value (**_attribute_value_**)))) (tag (start_tag (tag_name) (attribute (attribute_name) (quoted_attribute_value (**_attribute_value_**)))) (end_tag (tag_name))) (end_tag (tag_name)))
        (end_tag (tag_name))
    )
)

So even though the divs are nested, the way tree-sitter represents them is *not* nested. I would have assumed that something like `<div><div></div></div>` would produce `(start_tag ((start_tag) (end_tag)) (end_tag))` or something

I can't figure out how to work around this.

**Reply by sogaiu** [reply]
I found the output a bit hard to wrap my head around so here's a slightly reformatted version with colored delimiters for the sake of discussion:

![ts-tree-with-rainbow-parens-3](https://github.com/tree-sitter/tree-sitter/assets/33044872/30f8d72b-ae71-4b10-9e54-131213345008)

_Update_: reformatted a bit more
